File: CnA/as_rangesum/as_rangesum.cpp
#include<iostream>
#include <vector>
#include <algorithm>

int mergeSort(std::vector<long long int>& sum, int lower, int upper, int low, int high)
{
    if(high-low <= 1) return 0;
    int mid = (low+high)/2, m = mid, n = mid, count =0;
    count =mergeSort(sum,lower,upper,low,mid) +mergeSort(sum,lower,upper,mid,high);
    for(int i =low; i< mid; i++)
    {
        while(m < high && sum[m] - sum[i] < lower) m++;
        while(n < high && sum[n] - sum[i] <= upper) n++;
        count += n - m;
    }
    std::inplace_merge(sum.begin()+low, sum.begin()+mid, sum.begin()+high);
    return count;
}

int countRangeSum(std::vector<int>& nums, int lower, int upper) {
    int len = nums.size();
    std::vector<long long int> sum(len + 1, 0);
    for(int i =0; i< len; i++) sum[i+1] = sum[i]+nums[i];
    return mergeSort(sum, lower, upper, 0, len+1);
}

int main(){
	//get inputs
    int n, lower, upper;
    std::vector<int> nums;

    std::cin >> n>> lower >> upper;

    int tmp;
    for (int i=0; i<n; i++) {
        std::cin >> tmp;
        nums.push_back(tmp);
    }

    int count = countRangeSum(nums, lower, upper);

    
    std::cout<< count <<std::endl;

    return 0;
}

File: CnA/as_rangesum/as_rangesum.py
import sys

def countRangeSum(nums, lower, upper):
    n = len(nums)
    sums = []
    sums.append(0)
    for i in range(n):
        sums.append(sums[i] + nums[i])
    return countMergeSort(sums, 0, n+1, lower, upper)


def countMergeSort(sums, start, end, lower, upper):
    if (end-start <= 1):
        return 0
    
    mid = start + (end-start)//2
    j = k = t = mid
    
    count = countMergeSort(sums, start, mid, lower, upper) + countMergeSort(sums, mid, end, lower, upper)

    small_sum_store = []
    for i in range(start, mid):
        while (k < end) and (sums[k] - sums[i] < lower):
            k+=1
        while (j < end) and (sums[j] - sums[i] <= upper):
            j+=1
        while (t < end) and (sums[t] < sums[i]):
            small_sum_store.append(sums[t])
            t+=1
        small_sum_store.append(sums[i])
        count += (j-k)

    for i in range(t-start):
        sums[i + start] = small_sum_store[i]
    
    return count

if __name__ == '__main__':
	n, lower, upper = sys.stdin.readline().split(' ')
	n = int(n)
	lower = int(lower)
	upper = int(upper)
	
	nums = [int(x) for x in sys.stdin.readline().split(' ')]
	
	print(countRangeSum(nums, lower, upper))